File: 프로그래머스/2/340211. ［PCCP 기출문제］ 3번 ／ 충돌위험 찾기/［PCCP 기출문제］ 3번 ／ 충돌위험 찾기.java
import java.util.*;
import java.lang.*;

class Solution {
    public int solution(int[][] points, int[][] routes) {
        HashSet<String> set = new HashSet<>();
        HashSet<String> result = new HashSet<>();
        int robots = routes.length;
        
        for (int i = 0; i < robots; i++) {
            int start = routes[i][0] - 1;
            int startRow = points[start][0];
            int startCol = points[start][1];
            int cnt = 0;
            int sign;
            
            int routeCount = routes[i].length;
            String cur = startRow + "," + startCol + "," + cnt;
            
            if (!set.contains(cur)) {
                set.add(cur);    
            } else {
                result.add(cur);
            }
            
            for (int j = 1; j < routeCount; j++) {
                int to = routes[i][j] - 1;
                int toRow = points[to][0];
                int toCol = points[to][1];
                
                if (toRow > startRow) {
                    sign = 1;
                } else {
                    sign = -1;
                }
                
                int distance = Math.abs(toRow - startRow);
                
                for (int k = 0; k < distance; k++) {
                    cnt += 1;
                    startRow += sign;
                    
                    String next = startRow + "," + startCol + "," + cnt;
                    
                    if (!set.contains(next)) {
                        set.add(next);
                    } else {
                        result.add(next);
                    }
                    
                }
                
                if (toCol > startCol) {
                    sign = 1;
                } else {
                    sign = -1;
                }
                
                distance = Math.abs(toCol - startCol);
                
                for (int k = 0; k < distance; k++) {
                    cnt += 1;
                    startCol += sign;
                    String next = startRow + "," + startCol + "," + cnt;
                    
                    if (!set.contains(next)) {
                        set.add(next);
                    } else {
                        result.add(next);
                    }
                    
                }
            }
        }
        return result.size();
    }
}

File: 프로그래머스/2/340211. ［PCCP 기출문제］ 3번 ／ 충돌위험 찾기/［PCCP 기출문제］ 3번 ／ 충돌위험 찾기.py
def solution(points, routes):
    answer = 0
    sett = set()
    result = set()
    robots = len(routes)
    
    for i in range(robots):
        start = routes[i][0] - 1
        cur_routes = len(routes[i])
        cnt = 0
        start_r = points[start][0] # 시작 row
        start_c = points[start][1] # 시작 col
        
        if (start_r, start_c, cnt) not in sett:
            sett.add((start_r, start_c, cnt))
        else:
            result.add((start_r, start_c, cnt))
        
        for k in range(1, cur_routes):
            to = routes[i][k] - 1
            to_r = points[to][0] # 도착 row
            to_c = points[to][1] # 도착 col

            if to_r > start_r:
                sign = 1
            else:
                sign = -1

            for j in range(abs(to_r - start_r)):
                cnt += 1
                start_r += sign

                if (start_r, start_c, cnt) not in sett:
                    sett.add((start_r, start_c, cnt))
                else:
                    result.add((start_r, start_c, cnt))

            if to_c > start_c:
                sign = 1
            else:
                sign = -1

            for j in range(abs(to_c - start_c)):
                cnt += 1
                start_c += sign

                if (start_r, start_c, cnt) not in sett:
                    sett.add((start_r, start_c, cnt))
                else:
                    result.add((start_r, start_c, cnt))
                    
    return len(result)